caddy serve differently after reloading caddyfile failed due to a listen error

Hello, I came into this problem: when caddy reload failed due to a listen error, caddy serve differently to the same request and the reload function does't work. After the error, sometimes caddy serve with new caddyfile and sometimes with old caddyfile. **It seems like the old instance and new instance both exsits**.

Steps to reproduce the problem:

1. run caddy: `caddy -conf ./caddyfile -log stdout`. caddyfile is:

``` bash

http://test1.com:8080 {
    root /static/root
    log / stdout "{remote} [{when}] {uri}"
}

```

2. use `nc` to listen port 9999: 

```bash
nc -v -l 9999
```

3. change the caddy file and reload:  kill -s USR1 `pidof caddy`

``` bash

http://test1.com:8080 {
    root /static/root
    log / stdout "{remote} [{when}] {uri}"
}
## adding 2 sites
http://test2.com:8080 {
    root /static/root
    log / stdout "{remote} [{when}] {uri}"
}
## this cause a listen error
http://test3.com:9999 {
    root /static/root
    log / stdout "{remote} [{when}] {uri}"
}

```

4. Caddy will log a error 

```
[ERROR] Restart failed: Listen: listen tcp :9999: bind: address already in use
```

5. **PROBLEM1**. At this time, I think request to `test2.com:8080` should return 404, but:

``` bash

[x@verycool caddy]$ curl 127.0.0.1:8080/x.txt -H "host: test2.com"
404 Site test2.com is not served on this interface
[x@verycool caddy]$ curl 127.0.0.1:8080/x.txt -H "host: test2.com"
hello world
[x@verycool caddy]$ curl 127.0.0.1:8080/x.txt -H "host: test2.com"
404 Site test2.com is not served on this interface
[x@verycool caddy]$ curl 127.0.0.1:8080/x.txt -H "host: test2.com"
404 Site test2.com is not served on this interface
[x@verycool caddy]$ curl 127.0.0.1:8080/x.txt -H "host: test2.com"
hello world

```

6. stop `nc`, and reload again. Everything is ok.

7. change caddyfile:

``` bash

http://test1.com:8080 {
    root /static/root
    log / stdout "{remote} [{when}] {uri}"
}
http://test2.com:8080 {
    root /static/root
    ## change the access log format
    log / stdout "{remote} [{when}] {uri} ----"
}
http://test3.com:9999 {
    root /static/root
    log / stdout "{remote} [{when}] {uri}"
}

```

8. reload caddy again. Everything is ok and no error report.

9. **PROBLEM2**. At this time, I think request to `test2.com:8080` should produce log with `---.` But: 

``` bash

# cmd: for i in {1..1000}; do curl 127.0.0.1:8080/x.txt -H "host: test2.com"; done

# caddy log
127.0.0.1 [27/Feb/2019:18:07:35 +0800] /x.txt ----
127.0.0.1 [27/Feb/2019:18:07:35 +0800] /x.txt
127.0.0.1 [27/Feb/2019:18:07:35 +0800] /x.txt
127.0.0.1 [27/Feb/2019:18:07:35 +0800] /x.txt
127.0.0.1 [27/Feb/2019:18:07:35 +0800] /x.txt
127.0.0.1 [27/Feb/2019:18:07:35 +0800] /x.txt ----
127.0.0.1 [27/Feb/2019:18:07:35 +0800] /x.txt
127.0.0.1 [27/Feb/2019:18:07:35 +0800] /x.txt

```

Thanks for any help.

<!--
This template is for bug reports. The lack of a feature is not a bug; to request a feature, please switch templates.

Are you asking for help with using Caddy? Please ask on our forum: https://caddy.community

Please do not skip relevant questions; this will slow down the debugging process and your issue may be closed.
-->

## 1. Which version of Caddy are you using (`caddy -version`)?
<!-- If there is no version information, please paste commit SHA instead. -->

Caddy 0.11.4



## 2. What are you trying to do?
<!-- Please clearly describe what you are trying to do thoroughly enough so that a reader with no background information can repeat it. -->

change config and make it works by reloading config



## 3. What is your Caddyfile?

undef

<!-- Changing or hiding parts of your Caddyfile only slows things down and may result in your report being closed.
For more information, see https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222 -->
<!-- If you are unable to post this publicly, we offer private support: https://caddyserver.com/products/support -->


## 4. How did you run Caddy (give the full command and describe the execution environment)?
<!-- IMPORTANT: Please eliminate Docker, systemd, reverse proxies, upstream dependencies, caches, firewalls, and other unnecessary, external factors from your setup first. This will help prove that this is a bug in Caddy and not a misconfiguration of your environment. We may close issues that are too complex to replicate. Thank you! -->

caddy -conf caddyfile -log stdout


## 5. Please paste any relevant HTTP request(s) here.
<!-- Paste curl command, or full HTTP request including headers and body. You may skip this if the bug does not require HTTP requests. -->

curl 127.0.0.1:8080/x.txt -H "host: test2.com"


## 6. What did you expect to see?
<!-- Describe your expected results as precisely as possible. -->

1. 404
2. log with suffix `----`


## 7. What did you see instead (give full error messages and/or log)?
<!-- Please run Caddy with the -log flag, and use the log and errors directives as needed. DO NOT REDACT INFORMATION except for credentials. See https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222 -->

1. sometimes caddy return 200 and sometimes caddy return 404
2. sometimes caddy log `----` and sometimes not

## 8. Why is this a bug, and how do you think this should be fixed?
<!-- Help us understand why it is a bug; it is not always obvious. You can help us get this resolved faster by thinking about the problem and describing possible solutions! -->

1. If reload caddyfile failed, caddy still serve like before.
2. Reload should work if no errors report.

## 9. What are you doing to work around the problem in the meantime?
<!-- This can help others who encounter the same problem, until we implement a fix. -->

restart caddy


## 10. Please link to any related issues, pull requests, and/or discussion.
<!-- This can add crucial context to your report. -->

After reading the code, I find it may has someting to do with the time `s.Serve(ln)` called:

``` golang

// file: caddy.go
// function: func startServers(serverList []Server, inst *Instance, restartFds map[string]restartTriple) error 

    for _, s := range serverList {
        ....
        if ln == nil {
            ln, err = s.Listen()
            if err != nil {
                log.Printf("*** s.Listen failed: %s", err)
                return fmt.Errorf("Listen: %v", err)
            }
        }
        if pc == nil {
            pc, err = s.ListenPacket()
            if err != nil {
                return fmt.Errorf("ListenPacket: %v", err)
            }
        }

        //inst.wg.Add(2)
        //stopWg.Add(2)
        //func(s Server, ln net.Listener, pc net.PacketConn, inst *Instance) {
        //  go func() {
        //      defer func() {
        //          inst.wg.Done()
        //          stopWg.Done()
        //      }()
        //      errChan <- s.Serve(ln)
        //  }()

        //  go func() {
        //      defer func() {
        //          inst.wg.Done()
        //          stopWg.Done()
        //      }()
        //      errChan <- s.ServePacket(pc)
        //  }()
        //}(s, ln, pc, inst)

        inst.servers = append(inst.servers, ServerListener{server: s, listener: ln, packet: pc})
        log.Printf("append servers....")
    }

    // get s.Serve(ln) outside
    for _, s := range inst.servers {
        inst.wg.Add(2)
        stopWg.Add(2)
        func(s Server, ln net.Listener, pc net.PacketConn, inst *Instance) {
            go func() {
                defer func() {
                    inst.wg.Done()
                    stopWg.Done()
                }()
                errChan <- s.Serve(ln)
            }()

            go func() {
                defer func() {
                    inst.wg.Done()
                    stopWg.Done()
                }()
                errChan <- s.ServePacket(pc)
            }()
        }(s.server, s.listener, s.packet, inst)
    }

```

During the starting, one server listen failed will break the function, but the server before still takes place.

By change the code, the problem seems gone.

## Bonus: What do you use Caddy for? Why did you choose Caddy?
<!-- We'd like to know! -->

--- a/caddy.go
+++ b/caddy.go
@@ -781,6 +781,10 @@ func startServers(serverList []Server, inst *Instance, restartFds map[string]res
 			}
 		}
 
+		inst.servers = append(inst.servers, ServerListener{server: s, listener: ln, packet: pc})
+	}
+
+	for _, s := range inst.servers {
 		inst.wg.Add(2)
 		stopWg.Add(2)
 		func(s Server, ln net.Listener, pc net.PacketConn, inst *Instance) {
@@ -799,9 +803,7 @@ func startServers(serverList []Server, inst *Instance, restartFds map[string]res
 				}()
 				errChan <- s.ServePacket(pc)
 			}()
-		}(s, ln, pc, inst)
-
-		inst.servers = append(inst.servers, ServerListener{server: s, listener: ln, packet: pc})
+		}(s.server, s.listener, s.packet, inst)
 	}
 
 	// Log errors that may be returned from Serve() calls,

